### PR TITLE
Parts for DIYRobocarsFr hat

### PR DESCRIPTION
**Pull request by btrinite**
Add class to actuator to drive through Hat
Add new part for Hat as controller
Add relevant instances init and wiring in complete.py
Put throttle reverse filter as an option
Add our Jetson Nano camera config as comment
Add doc

### CODE REVIEW

**Review by obees — APPROVED** [review]
Great

File: donkeycar/parts/camera.py [hunk 1]
```diff
@@ -201,7 +201,7 @@ class CSICamera(BaseCamera):
     gstreamer init string from https://github.com/NVIDIA-AI-IOT/jetbot/blob/master/jetbot/camera.py
     '''
     def gstreamer_pipeline(self, capture_width=3280, capture_height=2464, output_width=224, output_height=224, framerate=21, flip_method=0) :   
-        return 'nvarguscamerasrc ! video/x-raw(memory:NVMM), width=%d, height=%d, format=(string)NV12, framerate=(fraction)%d/1 ! nvvidconv flip-method=%d ! nvvidconv ! video/x-raw, width=(int)%d, height=(int)%d, format=(string)BGRx ! videoconvert ! appsink' % (
+        return 'nvarguscamerasrc wbmode=3 sensor-mode=4 ! video/x-raw(memory:NVMM), width=%d, height=%d, format=(string)NV12, framerate=(fraction)%d/1 ! nvvidconv flip-method=%d ! nvvidconv ! video/x-raw, width=(int)%d, height=(int)%d, format=(string)BGRx ! videoconvert ! appsink' % (
```
Comment (DocGarbanzo) [inline donkeycar/parts/camera.py:204]: Can you break this super-overlong code line into pep8 acceptable line (I know it was already too long before :-)

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+import numpy as np
+from donkeycar.parts.actuator import RobocarsHat
+from donkeycar.utilities.logger import init_special_logger
+import socket
+import errno
+import sys
+import fcntl,os
+
+mylogger = init_special_logger ("Rx")
+mylogger.setLevel(logging.INFO)
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:15]: Similar comment as above.

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        self.ch3Feature = self._map_aux_feature (self.cfg.ROBOCARSHAT_CH3_FEATURE)
+        self.ch4Feature = self._map_aux_feature (self.cfg.ROBOCARSHAT_CH4_FEATURE)
+        if (self.ch3Feature == self.AUX_FEATURE_OUTPUT_STEERING_TRIM) or (self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_TRIM):
+            self.fixOutputSteeringTrim = 1500
+        if (self.ch3Feature == self.AUX_FEATURE_OUTPUT_STEERING_EXP) or (self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_EXP):
+            self.fixOutputSteering = 1500
+
+        if self.cfg.ROBOCARSHAT_THROTTLE_DISCRET != None:
+            self.discretesThrottle = np.arange(0.0,1.0001,1.0/len(self.cfg.ROBOCARSHAT_THROTTLE_DISCRET))
+            mylogger.info("CtrlIn Discrete throttle thresholds set to {}".format(self.discretesThrottle))
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:84]: Can you use f-strings?

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+        if self.cfg.ROBOCARSHAT_EMERGENCY_STOP == True:
+            self.emergemcyPort=9111
+            self.stop=False
+            self.server = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+            self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+            self.server.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+            fcntl.fcntl(self.server, fcntl.F_SETFL, os.O_NONBLOCK)
+            self.server.bind(("", self.emergemcyPort))
+            print("Listening emergency on port {}".format(self.emergemcyPort))
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:97]: Can you use logging instead of printing?

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+    def map_range(self, x, X_min, X_max, Y_min, Y_max):
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:99]: I don't like the fact that you are copying this method again, I see that it is already a bit messed up in the `utils.py` module, but we don't need the same thing 3x. Best would be to define the algorithm once (like here) and then have two versions, one which casts to `int` and another that rounds, both of these just calling the base algorithm. Or add an additional argument that does the rounding / casting on the result, if needed.

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+                    if params[3].isnumeric():
+                        self.inAux1 = self.map_range(int(params[3]),
+                            self.cfg.ROBOCARSHAT_PWM_IN_AUX_MIN, self.cfg.ROBOCARSHAT_PWM_IN_AUX_MAX,
+                            -1, 1)
+                    if params[4].isnumeric():
+                        self.inAux2 = self.map_range(int(params[4]),
+                            self.cfg.ROBOCARSHAT_PWM_IN_AUX_MIN, self.cfg.ROBOCARSHAT_PWM_IN_AUX_MAX,
+                            -1, 1)
+
+                    mylogger.debug("CtrlIn PWM {} {} {} {}".format(int(params[1]), int(params[2]), int(params[3]), int(params[4])))
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:153]: f-strings?

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+                    mylogger.debug("CtrlIn Std {} {} {} {}".format(self.inThrottle, self.inSteering, self.inAux1, self.inAux2))
+
+                if len(params) == 3 and int(params[0])==3 :
+                    if params[1].isnumeric():
+                        self.inThrottleIdle = int(params[1])
+                    if params[2].isnumeric():
+                        self.inSteeringIdle = int(params[2])
+                    mylogger.debug("CtrlIn Idle {} {} ".format(int(params[1]), int(params[2])))
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:161]: f-strings?

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+                
+                if len(params) == 3 and int(params[0])==2 :
+                    mylogger.debug("CtrlIn Sensors {} {} ".format(int(params[1]), int(params[2])))
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:164]: f-strings? - there are many more logger statements below, same applies to them.

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+                    if params[2].isnumeric():
+                        self.inSpeed = self.map_range(min(abs(int(params[2])),self.cfg.ROBOCARSHAT_ODOM_IN_MAX),
+                                    0, self.cfg.ROBOCARSHAT_ODOM_IN_MAX,
+                                1, 0)
+
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:169]: Too many blank lines (pep-8)

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+    def processAltModes(self):
+        self.recording=False
+        self.mode='user'
+        user_throttle = self.inThrottle
+        user_steering = self.inSteering
+
+        #Process Aux ch3
+        if self.ch3Feature == self.AUX_FEATURE_RECORDandPILOT :
+
+            if (self.inAux1<-0.5):
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:181]: Remove obsolete `()`, also below.

File: donkeycar/parts/robocars_hat_ctrl.py [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        return user_throttle, user_steering
+
+    def update(self):
+
+        while self.on:
+            start = datetime.now()
+            self.getCommand()
+            stop = datetime.now()
+            s = 0.01 - (stop - start).total_seconds()
+            if s > 0:
```
Comment (DocGarbanzo) [inline donkeycar/parts/robocars_hat_ctrl.py:328]: In general, you don't want to have any delay in your threaded loop, to give the part as little latency as possible. Is there any reason why you want to introduce a 10ms lag into the part?

**Pull request closed without merging**